pkgs/pkgs-lib/default.nix

```diff
@@ -7,5 +7,32 @@
   formats = import ./formats.nix {
     inherit lib pkgs;
   };
+
+  /*
+    Return attribute set of packages from overlay
+
+    Type: callOverlay :: (pkgs -> pkgs -> attrs) -> attrs
+
+    Example:
+      evalOverlay (final: prev: { a = 42; })
+      => { a = 42 }
+  */
+  callOverlay = overlay:
+    let
+      result = overlay (pkgs.extend overlay) pkgs;
```

Comment:
`pkgs.extend` creates a new instance of nixpkgs. And the user will only know about it if they read the code of `callOverlay`. At a minimum, this should be documented in the function documentation.

Reply:
As far as I know, it just applies a single overlay on top of `pkgs` which is super cheap. It does not evaluate everything over again.
I will double-check that this weekend.

An alternative approach would be to take care of `pkgs.pkgs` attribute and `pkgs.callPackage`, and it won't be clean and probably about as efficient as `pkgs.extend`

How would you approach this?

Reply:
Eh, decided that its not that difficult so I might as well do a benchmark right now:
![image](https://user-images.githubusercontent.com/23264966/178044863-e503cdda-7c6d-4e22-9326-74ef993b8b1c.png)

Test scenario:
```nix
{
  inputs.nixpkgs.url = github:gytis-ivaskevicius/nixpkgs/10e0ac7e0a764e64f967d78869671dda37001160;

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = nixpkgs.legacyPackages.${system};
    in
    {


      packages.${system} = {
        direct = pkgs.ruby;
        extends = (pkgs.extend (final: prev: { a = prev.ruby; })).a;
      };
    };
}
```

That's 17% increase in thunks. We probably could do better.

Reply:
Alright, i think i figured it out:
```
✦ ~/test-flake
❯ nix build .\#direct && jq .nrThunks stats.json
226531

✦ ~/test-flake
❯ nix build .\#liteBenchmark && jq .nrThunks stats.json
226534 # 3 thunks increase
```

Test scenario:
```nix
{
  inputs.nixpkgs.url = github:gytis-ivaskevicius/nixpkgs/10e0ac7e0a764e64f967d78869671dda37001160;

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = nixpkgs.legacyPackages.${system};
    in
    {
      packages.${system} = let

        liteOverlay = pkgs: fn: let
        newPkgs = fn newPkgs pkgs // {pkgs = newPkgs; callPackage = pkgs.lib.callPackageWith newPkgs; };
        in fn newPkgs pkgs;
        #in newPkgs;

      in {
        direct = pkgs.ruby;
        liteBenchmark = (liteOverlay pkgs (final: prev: { a = prev.ruby; })).a;
        extends = (pkgs.extend (final: prev: { a = prev.ruby; })).pkgs.pkgs.pkgs.callPackage ({a}: a){};
        lite = (liteOverlay pkgs (final: prev: { a = prev.ruby; })).pkgs.pkgs.pkgs.callPackage ({a}: a){};
      };
    };
}
```